File: app/src/main/java/se/gustavkarlsson/skylight/android/dagger/modules/AuroraReportModule.java
package se.gustavkarlsson.skylight.android.dagger.modules;

import android.net.ConnectivityManager;

import dagger.Binds;
import dagger.Module;
import dagger.Provides;
import dagger.Reusable;
import se.gustavkarlsson.skylight.android.BuildConfig;
import se.gustavkarlsson.skylight.android.background.providers.AsyncAddressProvider;
import se.gustavkarlsson.skylight.android.background.providers.AuroraFactorsProvider;
import se.gustavkarlsson.skylight.android.background.providers.AuroraReportProvider;
import se.gustavkarlsson.skylight.android.background.providers.LocationProvider;
import se.gustavkarlsson.skylight.android.background.providers.impl.AuroraReportProviderImpl;
import se.gustavkarlsson.skylight.android.background.providers.impl.DebugAuroraReportProvider;
import se.gustavkarlsson.skylight.android.background.providers.impl.aggregating_aurora_factors.AggregatingAuroraFactorsProvider;
import se.gustavkarlsson.skylight.android.settings.DebugSettings;

@Module
public abstract class AuroraReportModule {

	@Binds
	@Reusable
	abstract AuroraFactorsProvider bindAuroraFactorsProvider(AggregatingAuroraFactorsProvider provider);

	@Provides
	static AuroraReportProvider provideAuroraReportProvider(DebugSettings debugSettings, ConnectivityManager connectivityManager, LocationProvider locationProvider, AuroraFactorsProvider auroraFactorsProvider, AsyncAddressProvider asyncAddressProvider) {
		if (BuildConfig.DEBUG && debugSettings.isOverrideValues()) {
			return new DebugAuroraReportProvider(debugSettings);
		} else {
			return new AuroraReportProviderImpl(connectivityManager, locationProvider, auroraFactorsProvider, asyncAddressProvider);
		}
	}

}

File: app/src/debug/kotlin/se/gustavkarlsson/skylight/android/dagger/modules/AuroraReportModule.kt
package se.gustavkarlsson.skylight.android.dagger.modules

import android.content.Context
import com.f2prateek.rx.preferences2.RxSharedPreferences
import com.hadisatrio.optional.Optional
import com.jakewharton.rxrelay2.PublishRelay
import com.jakewharton.rxrelay2.Relay
import dagger.Module
import dagger.Provides
import dagger.Reusable
import io.reactivex.BackpressureStrategy
import io.reactivex.Flowable
import io.reactivex.Single
import io.reactivex.functions.Consumer
import org.threeten.bp.Instant
import se.gustavkarlsson.skylight.android.entities.*
import se.gustavkarlsson.skylight.android.services.DebugSettings
import se.gustavkarlsson.skylight.android.services.Streamable
import se.gustavkarlsson.skylight.android.services.providers.*
import se.gustavkarlsson.skylight.android.services_impl.RxPreferencesDebugSettings
import se.gustavkarlsson.skylight.android.services_impl.providers.CombiningAuroraReportProvider
import se.gustavkarlsson.skylight.android.services_impl.providers.DebugAuroraReportProvider
import se.gustavkarlsson.skylight.android.services_impl.streamables.CombiningAuroraReportStreamable
import se.gustavkarlsson.skylight.android.services_impl.streamables.DebugAuroraReportStreamable
import javax.inject.Singleton

@Module
class AuroraReportModule {

	@Provides
	@Reusable
	fun provideDebugSettings(
		rxSharedPreferences: RxSharedPreferences,
		context: Context
	): DebugSettings = RxPreferencesDebugSettings(context, rxSharedPreferences)

	@Provides
	@Reusable
	fun provideAuroraReportProvider(
		timeProvider: TimeProvider,
		locationProvider: LocationProvider,
		locationNameProvider: LocationNameProvider,
		darknessProvider: DarknessProvider,
		geomagLocationProvider: GeomagLocationProvider,
		kpIndexProvider: KpIndexProvider,
		visibilityProvider: VisibilityProvider,
		debugSettings: DebugSettings
	): AuroraReportProvider {
		val realProvider = CombiningAuroraReportProvider(
			timeProvider,
			locationProvider,
			locationNameProvider,
			darknessProvider,
			geomagLocationProvider,
			kpIndexProvider,
			visibilityProvider
		)
		return DebugAuroraReportProvider(realProvider, debugSettings, timeProvider)
	}

	@Provides
	@Reusable
	fun provideAuroraReportSingle(
		auroraReportProvider: AuroraReportProvider,
		relay: Relay<AuroraReport>
	): Single<AuroraReport> {
		return auroraReportProvider.get()
			.doOnSuccess(relay)
	}

	@Provides
	@Singleton
	fun provideAuroraReportRelay(
	): Relay<AuroraReport> {
		return PublishRelay.create()
	}

	@Provides
	@Reusable
	fun provideAuroraReportConsumer(
		relay: Relay<AuroraReport>
	): Consumer<AuroraReport> {
		return relay
	}

	@Provides
	@Reusable
	fun provideAuroraReportStreamable(
		now: Single<Instant>,
		locationNames: Flowable<Optional<String>>,
		kpIndexes: Flowable<KpIndex>,
		geomagLocations: Flowable<GeomagLocation>,
		darknesses: Flowable<Darkness>,
		visibilities: Flowable<Visibility>,
		debugSettings: DebugSettings,
		relay: Relay<AuroraReport>
	): Streamable<AuroraReport> {
		val realStreamable = CombiningAuroraReportStreamable(
			now,
			locationNames,
			kpIndexes,
			geomagLocations,
			darknesses,
			visibilities,
			relay.toFlowable(BackpressureStrategy.LATEST)
		)
		return DebugAuroraReportStreamable(realStreamable, debugSettings, now)
	}

	@Provides
	@Reusable
	fun provideAuroraReportFlowable(
		streamable: Streamable<AuroraReport>
	): Flowable<AuroraReport> = streamable.stream
		.startWith(AuroraReport.empty)
		.replay(1)
		.refCount()
}
